### PR TITLE
Fixed the problem of disbinding order between epoll and corresponding socket/group

### LINKED ISSUE #957: Incorrect order of disbinding the being-removed socket from epoll-eid
The `srt_epoll_remove_usock` function should disbind the socket being removed and the epoll container, which means two things:
1. Remove the socket from the epoll container through the epoll system
2. Remove the `eid` designating this epoll from the eid container of that socket

Removal in #2, done by `CUDT::removeEPoll`, is also doing update-to-false of the IN/OUT events first so that this socket no longer reports any readiness. The problem is that in #1 this socket is first removed from the epoll container, so the `update_events` function, when trying to refer to that epoll container, will fail to find the associated watch object, as it has been just removed in #1. 

What needs to be done: research as to whether removal of the readiness events is really required in this case, try if this `CUDT::removeEPoll` function is being called also in different places, and if the event resetting is needed, maybe it can be shifted to another place.

This problem doesn't cause any real problem here, just it falls under a condition which should not occur.

### PR DESCRIPTION
**Pull request by ethouris**
Fixes #957

The problem:

Nothing more than appearing IPE message (although meaning a potential more serious problem in future), when the `update_event` function was called, which normally happens when the SRT entity (socket or group) becomes ready for particular event.

The structure for that is such that the epoll container contains the list of SRT entities by their IDs so that it can be quickly tell to the subsribers when they become (or are) ready. However every SRT entity also has its own container that contains the list of epoll container IDs (EID) that are subscribed to them (this time, for whatever kind of events). This is kept because the epoll readiness state change function, `update_event`, is running over this list first so that it only updates those epoll containers that have subscribed to that entity. This however constitutes a two-way binding, which means that dis-binding them must involve also both references - in the epoll container and in the entity itself.

The problem was twofold:

1. When unsubscribing, there was also undertaken a step to remove first readiness state from given epoll container. This was, however, done after first unsubscribing from the container, so clearing events was wrong, as it was done against the EID from which's container the entity was first unsubscribed. OTOH it still looks beyond the reason, as when unsubscribing, the notice is removed as well - although things have changed around here since the refactoring of epoll system, as well as readiness is marked not only in the epoll. However, clear-after-unsubscribed has generated an IPE message.

2. When closing a socket it was first set the ERR event on it, and it was intended to leave it subscribed for a little while, as the socket cleanup was about to be done later in the GC thread, so the socket should be able to finish its job. It didn't have much of a reason though stating that the next instruction was forcefully unsubscribing it from the epoll container. Not sure what solution should really be done here, but if we state that this unsubscribing was the correct way of doing things, then the corresponding EIDs from which unsubscribing happened, should be also complemented with removal of all these EIDs from the entity's registry - as otherwise it may contain in itself EIDs that it believes it is subscribed to, but this not only isn't true, but the corresponding EIDs may even no longer exist. The matter of whether the application has at least any chance to learn about this socket's destruction from epoll system is believed to be solved by the fact that the waiting function will be woken up at the moment when the update happens, that sets the ERR event, so at the moment when the subscribtion is deleted - and also corresponding EIDs removed from the entity's registry - this socket no longer exists anyway.

By the way there have been some more logging added to the functions around the described problem as well as some logging added to the test case. Also, compat-shortcut functions, `add_usock` and `remove_usock` have been deleted. These are function used only internally.

### CODE REVIEW

**Comment by jeandube** [comment]
I had a storm of these error with my test app and master head (1.4.1+)
```08:38:46.728456 mxptool: [2]srt: Connection established to: 10.65.10.242:9942 (v1.3.1)
08:38:46.728468 mxptool: update_events(epoll.cpp:839): /UEpoll*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214053
SRT version 10401 -> 10301
08:38:46.728640 mxptool: update_events(epoll.cpp:839): /App:Rx*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054
08:38:46.741464 mxptool: update_events(epoll.cpp:839): /UEpoll*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054
08:38:46.741614 mxptool: update_events(epoll.cpp:839): /App:Rx*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054
08:38:46.744376 mxptool: update_events(epoll.cpp:839): /UEpoll*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054
08:38:46.744543 mxptool: update_events(epoll.cpp:839): /App:Rx*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054
08:38:46.760405 mxptool: update_events(epoll.cpp:839): /UEpoll*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054
08:38:46.760574 mxptool: update_events(epoll.cpp:839): /App:Rx*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054
08:38:46.762453 mxptool: update_events(epoll.cpp:839): /UEpoll*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054
08:38:46.762599 mxptool: update_events(epoll.cpp:839): /App:Rx*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054
08:38:46.776522 mxptool: update_events(epoll.cpp:839): /UEpoll*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054
08:38:46.776699 mxptool: update_events(epoll.cpp:839): /App:Rx*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054
08:38:46.784073 mxptool: update_events(epoll.cpp:839): /UEpoll*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054
08:38:46.784219 mxptool: update_events(epoll.cpp:839): /App:Rx*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @489214054```

So I tried this PR and error log is still present but its rate is way lower.
```09:10:28.197110 mxptool: [2]srt: Connection established to: 10.65.10.242:9942 (v1.3.1)
SRT version 10401 -> 10301
[Thread 0x7fffed7fa700 (LWP 13876) exited]
[Thread 0x7fffedffb700 (LWP 13875) exited]
09:10:32.850109 mxptool: update_events(epoll.cpp:864): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @184136526
09:10:33.585386 mxptool: update_events(epoll.cpp:864): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @184136526
09:10:34.339937 mxptool: update_events(epoll.cpp:864): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @184136526
09:10:34.500521 mxptool: update_events(epoll.cpp:864): /App:Rx*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @184136526
09:10:40.557083 mxptool: update_events(epoll.cpp:864): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @184136526```

In both cases the app is working fine.

File: test/test_epoll.cpp [hunk 1]
```diff
@@ -417,7 +418,8 @@ TEST(CEPoll, HandleEpollEvent2)
 
     try
     {
-        EXPECT_EQ(epoll.remove_usock(epoll_id, client_sock), 0);
+		int no_events = 0;
```
Comment (maxsharabayko) [inline test/test_epoll.cpp]: Alignment
```suggestion
        int no_events = 0;
```

File: test/test_epoll.cpp [hunk 2]
```diff
@@ -468,7 +470,8 @@ TEST(CEPoll, HandleEpollNoEvent)
...
```
Comment (maxsharabayko) [inline test/test_epoll.cpp]: Alignment

File: test/test_epoll.cpp [hunk 3]
```diff
@@ -534,7 +537,8 @@ TEST(CEPoll, ThreadedUpdate)
...
```
Comment (maxsharabayko) [inline test/test_epoll.cpp]: Alignment

File: test/test_enforced_encryption.cpp [hunk 4]
```diff
@@ -382,19 +432,15 @@ class TestEnforcedEncryption
                     }
                 }
             }
-        });
 
-        const int connect_ret = srt_connect(m_caller_socket, psa, sizeof sa);
-        EXPECT_EQ(connect_ret, expect.connect_ret);
```
Comment (maxsharabayko) [inline test/test_enforced_encryption.cpp:388]: Moving this part up looks like reviving the old version of the test with higher likelihood of failing due to missing socket closure.

Comment (ethouris) [reply test/test_enforced_encryption.cpp:388]: I think this whole change for this file deserves to get into a different PR. Even though this test was being revised during the works on this.

**Comment by boxerab** [comment]
:raised_hands:

**Comment by jeandube** [comment]
I still reproduce this problem for an listen accepted socket with commit 92abcd6 (June 19), streamID and listen-callback, no group, no closing, stable rx-only streaming socket.
```07:15:37.986786 mxptool: worker_ProcessConnectionRequest(queue.cpp:1358): /SRT:RcvQ:w1.N:SRT.c: Listener managed the connection request from: ::ffff:10.65.10.234:9931 result:waveahand
07:15:37.986799 mxptool: [1:L]srt: sock[14ab458e] accepting newsock[14ab458b]
07:15:37.986883 mxptool: [5]srt: create
07:15:37.986912 mxptool: [5]srt: sock[0->14ab458b] listening->connected
07:15:37.986942 mxptool: [5]srt: Connection established from: [::ffff:10.65.10.234]:9931 (v1.4.1)
07:15:37.986973 mxptool: [1:L]srt: new connection clone: 5
07:15:37.986992 mxptool: [1:L]srt: new connection event: 5
cxFindFromResID([ 0],tx,dubm4e/bnc1/4mbps)
[ 1] 0w
[ 0] rx
[New Thread 0x7fffedffb700 (LWP 8045)]
[New Thread 0x7fffeeffd700 (LWP 8046)]
recv from [::ffff:10.65.10.234]:9931
07:15:38.917323 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:15:39.284391 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:15:39.668118 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:15:45.809965 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:15:49.813544 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:15:55.819103 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:16:07.830607 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:16:11.835853 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:16:13.837127 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:17:17.901267 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:18:40.516294 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:19:06.009997 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:20:24.020303 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:21:30.155059 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:21:58.183192 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:22:30.215000 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:22:44.229374 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:23:04.249476 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:23:34.244437 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:32:31.000889 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:33:04.853748 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:37:17.106826 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:41:21.352092 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:41:23.354805 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:41:25.356042 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:41:33.364882 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:41:37.368141 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:41:43.374484 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:41:45.376153 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:41:47.378179 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:41:51.382651 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779
07:41:55.386102 mxptool: update_events(epoll.cpp:858): /SRT:TsbPd*E:SRT.d: epoll/update: IPE: update struck EID 1 which is NOT SUBSCRIBED to @346768779```

**Pull request merged**